pubsublite/admin.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package pubsublite
+
+import (
+	"context"
+
+	"google.golang.org/api/option"
+	"google.golang.org/api/option/internaloption"
+
+	vkit "cloud.google.com/go/pubsublite/apiv1"
+	pb "google.golang.org/genproto/googleapis/cloud/pubsublite/v1"
+)
+
+// AdminClient provides admin operations for Google Pub/Sub Lite resources
+// within a Google Cloud region. An AdminClient may be shared by multiple
+// goroutines.
+type AdminClient struct {
+	admin *vkit.AdminClient
+}
+
+// NewAdminClient creates a new Cloud Pub/Sub Lite client to perform admin
+// operations for resources within a given region.
+// See https://cloud.google.com/pubsub/lite/docs/locations for the list of
+// regions and zones where Google Pub/Sub Lite is available.
+func NewAdminClient(ctx context.Context, region string, opts ...option.ClientOption) (*AdminClient, error) {
+	if err := validateRegion(region); err != nil {
+		return nil, err
+	}
+	options := []option.ClientOption{internaloption.WithDefaultEndpoint(region + "-pubsublite.googleapis.com:443")}
+	options = append(options, opts...)
+	admin, err := vkit.NewAdminClient(ctx, options...)
+	if err != nil {
+		return nil, err
+	}
+	return &AdminClient{admin: admin}, nil
+}
+
+// CreateTopic creates a new topic from the given config.
+func (ac *AdminClient) CreateTopic(ctx context.Context, config TopicConfig) (*TopicConfig, error) {
+	req := &pb.CreateTopicRequest{
+		Parent:  config.Name.location().String(),
+		Topic:   config.toProto(),
+		TopicId: config.Name.TopicID,
+	}
+	topicpb, err := ac.admin.CreateTopic(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return protoToTopicConfig(topicpb)
+}
+
+// UpdateTopic updates an existing topic from the given config and returns the
+// new topic config.
+func (ac *AdminClient) UpdateTopic(ctx context.Context, config TopicConfigToUpdate) (*TopicConfig, error) {
+	topicpb, err := ac.admin.UpdateTopic(ctx, config.toUpdateRequest())
+	if err != nil {
+		return nil, err
+	}
+	return protoToTopicConfig(topicpb)
+}
+
+// DeleteTopic deletes a topic.
+func (ac *AdminClient) DeleteTopic(ctx context.Context, topic TopicPath) error {
```

Comment:
Here, and with other methods, XXXPath does not seems like a meaningful name. It makes sense if you know that it is used to build a resource name, but if you don't know that the name is confusing in my opinion. I wonder if there is a way to refactor this so that all you need to pass to the method is a  context and `TopicID`.

Reply:
The topic is uniquely identified by its {project, zone, topic id}. I didn't store the project and zone in the client because we can have cross-project subscriptions, where the topic and subscription resources are not necessarily owned by the same project. So having a default project set for the client could cause some confusion for, e.g. Create/Update Subscription (i.e. is the default project used for the topic or the subscription?).

I also saw that [pubsub.Client](https://godoc.org/cloud.google.com/go/pubsub#pkg-index) had evolved to support multiple variants Topic/TopicInProject, Subscription/SubscriptionInProject, so wanted to avoid this, as it seems multi-project admin does occur.

Similar remarks for zone, as the admin server is regional. Users can use the same client to manage all resources within the region.

For these reasons, I thought it would be best to give the client more flexibility with projects and zones; avoiding the multiple variants that pubsub.Client has.

Reply:
I forgot to mention: we have planned regional replication for topics (b/171412217), so the admin client could end up dealing with multiple zones within the same region, so storing a default zone could cause confusion and/or proliferation of method variants.